isochrone/src/main/java/com/graphhopper/isochrone/algorithm/ShortestPathTree.java

```diff
@@ -187,7 +194,11 @@ private double getExploreValue(IsoLabel label) {
 
     @Override
     protected boolean finished() {
-        return queueByZ.isEmpty() || getExploreValue(queueByZ.peek()) >= limit;
+        while (queueByZ.peek() != null && queueByZ.peek().deleted)
+            queueByZ.poll();
+        if (queueByZ.peek() == null)
+            return true;
+        return getExploreValue(queueByZ.peek()) >= limit;
```

Comment:
```java
while (!queueByZ.isEmpty()) {
    IsoLabel head = queueByZ.poll();
    if (!head.deleted) {
        return getExploreValue(head) >= limit;
    }
}
return true;
```

Depending on the size of the underlying array it would probably be faster to sort it once and iterate over the array.

Reply:
The Java PriorityQueue is an extreme fast implementation (also backed on an array) and so it is incredible hard to beat it even if alternatives are theoretically faster.

Reply:
It would also allow to use a more finegrained Comparator and sort by deleted+z